### ISSUE TITLE
Scrap Mechanic (387990)

### ISSUE THREAD

**Issue by Scrumplex** [ValveSoftware/Proton#2165]
# Compatibility Report
- Name of the game with compatibility issues: Scrap Mechanic
- Steam AppID of the game: 387990

## System Information
- GPU: GTX 1060 6GB
- Driver/LLVM version: nvidia 415.25
- Kernel version: 4.19.12-zen1-1-zen
- Link to full system information report as [Gist](https://gist.github.com/): https://gist.github.com/Scrumplex/5f37aeae37ecd876952b4f11e7742d15
- Proton version: 3.16-6 Beta

## I confirm:
- [x] that I haven't found an existing compatibility report for this game.
- [x] that I have checked whether there are updates for my system available.

[steam-387990.log](https://github.com/ValveSoftware/Proton/files/2717299/steam-387990.log)

## Symptoms 
The game does not start. It only shows a black screen. It used to work before one of the latest updates of the game with the WINED3D flag enabled.

## Reproduction
Just open the game.

**Comment by TaiTair** [comment]
Game runs fine with kernel 4.20, driver 418.56, proton 4.2.5 and WINED3D flag.

Wish it worked with DXVK though. With DXVK fails to create graphics device.

**Comment by TaiTair** [comment]
If I run it without WINED3D it crashes and produces logs under ScrapMechanic/Logs.
The game log shows this:
```
13:11:56 (1) [Default] Initializing Logger
13:11:56 (1) [System] Operating system: XP Service pack 3
13:11:56 (1) [System] Permissions for C:/users/steamuser/Application Data - Others write | Others read | Group write | Group read | Owner write | Owner read
13:11:56 (1) [System] Permissions for C:/users/steamuser/Local Settings/Application Data - Others write | Others read | Group write | Group read | Owner write | Owner read
13:11:56 (1) [System] Permissions for .. - Others write | Others read | Group write | Group read | Owner write | Owner read
13:11:56 (1) [Default] Initializing ResourceManager
13:11:56 (1) ERROR: d:\jenkinsworkspace\scrapmechanic\contraptioncommon\resources\resourcemanager.cpp:89 $GAME_DATA/Objects/Mesh/fittings/obj_fittings_fusebox.mesh is marked for upgrade, but the file exists. Ignoring upgrade pair
13:11:56 (1) [System] GPU: GeForce GTX 1050
13:11:56 (1) [System] Dedicated Video Memory: 3072 Mb
13:11:56 (1) [System] Shared System Memory: 3072 Mb
13:11:56 (1) [System] CPU: Intel(R) Core(TM) i7-7700HQ CPU @ 2.80GHz
13:11:56 (1) [System] Total RAM: 15921 Mb
13:11:56 (1) [System] Available RAM: 12463 Mb
13:11:57 (1) [System] Supported feature level: 11.0
13:11:57 (1) [Render] Dx11 device created
13:11:58 (1) [Default] Initializing GUI
13:11:58 (1) [Default] Creating MyGUI DX11 platform
13:11:59 (1) ERROR: d:\jenkinsworkspace\scrapmechanic\scrapmechanic\contraption.cpp:802 Bugsplat!
```

Which confirms DXVK is the issue.
There is a second log for MyGUI:

```
13:11:58  |  Platform  |  Info  |  * Initialise: RenderManager  |  z:\deps\vc141\mygui\platforms\directx11\directx11platform\src\mygui_directx11rendermanager.cpp  |  85
13:11:58  |  Platform  |  Critical  |  RenderManager 0 viewport sets  |  z:\deps\vc141\mygui\platforms\directx11\directx11platform\src\mygui_directx11rendermanager.cpp  |  130
```

It looks like it fails to create a Graphics device because it is using VC141.

Edit: forgot to change Windows version back to 7 with winecfg. Output remains the same with 7 though.

**Comment by TaiTair** [comment]
Update: Fixed DXVK, Launched steam with terminal, moved game back to NTFS partition, verified installation files. Game launches correctly from NTFS partition.
DXVK fix might be irrelevant for this since I use PROTON_USE_WINED3D flag for this game.
I think what fixed it was:

1. Setting up the symbolic link for compatdata on the NTFS partition.
2. Verifying files after moving installation folder.

**Comment by MadByteDE** [comment]
[DXVK receives a fix](https://github.com/doitsujin/dxvk/pull/1117) for the blackscreen / crash on startup.. You can built DXVK from source or use the following link to get it before it get's updated in a future Proton version. The archive contains setup instructions if needed.
https://drive.google.com/open?id=1Q8ymLwk2tB2zWQOh5_RDVllMeC_duJ5N

**Comment by TaiTair** [comment]
I will wait for proton to be updated and then I'll try. I'm assuming this will probably be in proton 5.0.
Just read the issue on the DXVK git. That looks very promising. I'm excited to try it out.

**Comment by Friz64** [comment]
Scrap Mechanic is still broken for AMD/Intel Mesa users because of another bug where the game crashes with this error:
![image](https://user-images.githubusercontent.com/20155479/61183067-8fa95a80-a63c-11e9-93a5-f92559e311af.png)
This is the log that gets created:
```
13:34:12 (1) [Default] Initializing Logger
13:34:12 (1) [System] Operating system: Win 7 Service pack 1
13:34:12 (1) [System] Permissions for C:/users/steamuser/Application Data - Others write | Others read | Group write | Group read | Owner write | Owner read
13:34:12 (1) [System] Permissions for C:/users/steamuser/Local Settings/Application Data - Others write | Others read | Group write | Group read | Owner write | Owner read
13:34:12 (1) [System] Permissions for .. - Others write | Others read | Group write | Group read | Owner write | Owner read
13:34:12 (1) [Default] Initializing ResourceManager
13:34:12 (1) ERROR: d:\jenkinsworkspace\scrapmechanic\contraptioncommon\resources\resourcemanager.cpp:89 $GAME_DATA/Objects/Mesh/fittings/obj_fittings_fusebox.mesh is marked for upgrade, but the file exists. Ignoring upgrade pair
13:34:12 (1) ERROR: d:\jenkinsworkspace\scrapmechanic\contraptioncommon\renderer\dx11\d3d11_device.cpp:127 DirectX Error "CreateDXGIFactory" failed. Error code: 0x80004005

13:34:12 (1) [Render] ERROR: d:\jenkinsworkspace\scrapmechanic\contraptioncommon\renderer\dx11\d3d11_renderer.cpp:59 Adapter search failed
Shutting down...
```

This has been confirmed by a few AMD/Intel Mesa users on [protondb](https://www.protondb.com/app/387990).

**Comment by Friz64** [comment]
I have opened an DXVK issue for this. https://github.com/doitsujin/dxvk/issues/1129

**Comment by Friz64** [comment]
Update: This patch also works just fine using AMD/Intel Mesa, it just requires the 32-bit vulkan drivers to be installed.

**Comment by Mhowser** [comment]
Hi, there is a new issue with Scrap Mechanic using Proton 5.0-6. The game starts fine but the cursor disappears when it goes into the game window and nothing is selectable/clickable, i cant even use TAB to navigate. The steam overlay is functional.
I'm using amd cpu and amd gpu on Zorin OS 15 (Ubuntu 18.04)
Here is more GPU info:
```
Graphics:  Card: Advanced Micro Devices [AMD/ATI] Ellesmere [Radeon RX 470/480/570/570X/580/580X]
           Display Server: wayland (X.Org 1.19.6 )
           drivers: fbdev,ati (unloaded: modesetting,vesa,radeon)
           Resolution: 1280x1024@59.89hz, 1920x1080@59.96hz
           OpenGL: renderer: Radeon RX 580 Series (POLARIS10, DRM 3.36.0, 5.5.0-050500-generic, LLVM 10.0.0)
           version: 4.6 Mesa 20.0.5 - kisak-mesa PPA
```

**Comment by kisak-valve** [comment]
Hello @Mhowser, please add [`PROTON_LOG=1 %command%`](https://github.com/ValveSoftware/Proton#runtime-config-options) to the game's launch options and drag and drop the generated $HOME/steam-$APPID.log into the comment box. Also, what's the most recent version of Proton which doesn't show the regression?

**Comment by Mhowser** [comment]
Hi, sorry for taking so long to respond, that bug was fixed by switching over to Xorg instead of using Wayland.

There is now a new bug with BETA 0.4.0 that was just released, the game crashes everytime I try to create or load a world, be it survival or creative.
I dont see anything of interest in the game log but here it is; 
[game-20200507-103025.log](https://github.com/ValveSoftware/Proton/files/4594778/game-20200507-103025.log)

Please let me know if you need more info just tell me how to get it.

**Comment by Friz64** [comment]
```
20:03:03 (1) [Default] Opening save file: 'C:/users/steamuser/Application Data/Axolot Games/Scrap Mechanic/User/User_76561198242564406/Save/Survival/test1.db'
20:03:07 (1) ERROR: Z:\Jenkins\workspace\sm\ScrapMechanic\Contraption.cpp:856 Bugsplat!
```
Same here, this doesn't look very fixable for the end user unfortunately. The best thing to do now is to contact the developers and hope that they try to fix the issue. I have already sent them a tweet.

**Comment by TaiTair** [comment]
~~It look like they switched Scrap Mechanic to DX12 so it's running on VKD3D now.~~  
My framerate limiters don't work anymore. Game doesn't run in openGL anymore (with PROTON_USE_WINED3D=1)
Haven't tried launching a game yet since I'm trying to disable DX12.

**Comment by Friz64** [comment]
> It look like they switched Scrap Mechanic to DX12 so it's running on VKD3D now. My framerate limiters don't work anymore. Game doesn't run in openGL anymore (with PROTON_USE_WINED3D=1)
> Haven't tried launching a game yet since I'm trying to disable DX12.

Where did you get that idea from? For me it still seems like it's running on DX11.
```20:02:02 (1) [Render] Dx11 device created```

**Comment by TaiTair** [comment]
> > It look like they switched Scrap Mechanic to DX12 so it's running on VKD3D now. My framerate limiters don't work anymore. Game doesn't run in openGL anymore (with PROTON_USE_WINED3D=1)
> > Haven't tried launching a game yet since I'm trying to disable DX12.
> 
> Where did you get that idea from? For me it still seems like it's running on DX11.
> `20:02:02 (1) [Render] Dx11 device created`

Ok my bad. It was an assumption because VKGHL was not working and because I couldn't get it to run on WINED3D. Checked the logs and you are right. It's DX11.
Bunch of fishy stuff going on though:
`14458.162:00bc:00c0:trace:loaddll:load_native_dll Loaded L"C:\\windows\\system32\\dxvk_config.dll" at 0x69240000: native
warn:  D3D11CoreCreateDevice: Adapter is not a DXVK adapter                                                                
info:  Game: ScrapMechanic.exe                                                                                                                                            
info:  DXVK: v1.6-69-g5f4f92002              `


`14463.404:00bc:00fc:err:ntdll:RtlpWaitForCriticalSection section 0x1672208 "?" wait timed out in thread 00fc, blocked by 00f8, retrying (60 sec)
14463.404:00bc:0104:err:ntdll:RtlpWaitForCriticalSection section 0x1672208 "?" wait timed out in thread 0104, blocked by 00f8, retrying (60 sec)
14463.404:00bc:00f8:err:ntdll:RtlpWaitForCriticalSection section 0x7bebf780 "../../../../proton/wine/dlls/ntdll/loader.c: loader_section" wait timed out in thread 00f8, blocked by 0104, retrying (60 sec)
14463.404:00bc:0108:err:ntdll:RtlpWaitForCriticalSection section 0x7bebf780 "../../../../proton/wine/dlls/ntdll/loader.c: loader_section" wait timed out in thread 0108, blocked by 0104, retrying (60 sec)
14463.404:00bc:010c:err:ntdll:RtlpWaitForCriticalSection section 0x7bebf780 "../../../../proton/wine/dlls/ntdll/loader.c: loader_section" wait timed out in thread 010c, blocked by 0104, retrying (60 sec)
`

Half the time, game doesn't even startup.
[steam-387990.log](https://github.com/ValveSoftware/Proton/files/4595209/steam-387990.log)

**Comment by Mhowser** [comment]
@TaiTair what are your launch options?

**Comment by Friz64** [comment]
I don't know much about Proton internals but i don't think they are actually the problem. The real reason of the crash seems to be some kind of assertion *in Scrap Mechanic* or something, not anything we can control i think.

**Comment by TaiTair** [comment]
On WINED3D
`14849.034:00bc:00c0:fixme:d3d:wined3d_swapchain_init The application requested more than one back buffer, this is not properly supported.
Please configure the application to use double buffering (1 back buffer) if possible.
`
[steam-387990.log](https://github.com/ValveSoftware/Proton/files/4595226/steam-387990.log)



> @TaiTair what are your launch options?

I tried a few: wined3d, nod3d12, nod3d11, nod3d10, noesync, nofsync, seccomp 

Seems to be on Scrap Mechanic end.

**Comment by UltraBlackLinux** [comment]
Do u guys that are having that problem have an AMD Graphics card? 
this may actually be he problem since all the people with the lauch problems have them.

**Comment by Mhowser** [comment]
> Do u guys that are having that problem have an AMD Graphics card?
> this may actually be he problem since all the people with the lauch problems have them.

I have the Radeon RX 580

**Comment by Friz64** [comment]
I have a Radeon RX 580 too, but i don't think it's the cause.

**Comment by TaiTair** [comment]
No. Nvidia 1050 here.

**Comment by TaiTair** [comment]
Ok so I tried a few more things just to make sure.
Made a dxvk.conf for Scrap Mechanic:
-Tried disabling nvapihack to make my GPU run as an Nvidia GPU, no dice.
-Tried limiting DX feature level to 10_1, 9_3. No dice.
-Tried setting dxgi.syncInterval to 0.5 to force 30 fps. No dice. (Ran set to 1 though)

I've noticed a couple things during my tests:
~~-Adding DXVK_HUD=1 to launch parameters does not enable the DXVK HUD, it can only be enabled from dxvk.conf for this game.~~
-This is the only game right now in my library that does not use the layers I set in vkconfig. VKGHL still works with my other games. 
This makes me think the game is actually pulling off some weird rendering tricks.

I also tried launching the game in creative instead of survival. No difference.

**Comment by Friz64** [comment]
> -Adding DXVK_HUD=1 to launch parameters does not enable the DXVK HUD, it can only be enabled from dxvk.conf for this game.

I use `DXVK_HUD=full %command` as a start command just fine. Again, i still doubt that the crash has anything to do with rendering and believe it has more to do with some other incompatibility.

**Comment by TaiTair** [comment]
> > -Adding DXVK_HUD=1 to launch parameters does not enable the DXVK HUD, it can only be enabled from dxvk.conf for this game.
> 
> I use `DXVK_HUD=full %command` as a start command just fine. Again, i still doubt that the crash has anything to do with rendering and believe it has more to do with some other incompatibility.

Ok I guess I forgot the %command% because it does work.

That doesn't explain why the game doesn't use my vulkan layers though. It's not loading my configuration so it's doing weird stuff. Who knows what else it's not loading other than vulkan layers.

**Comment by UltraBlackLinux** [comment]
even though steam and SM recieved updates tday, nothing has been fixed about this.

**Comment by wozeparrot** [comment]
The game seems to run fine with steam installed under wine.

**Comment by Friz64** [comment]
> The game seems to run fine with steam installed under wine.

Meaning that you can also join worlds?

**Comment by wozeparrot** [comment]
Yes you can create a world and it works. The only thing that does not work is multiplayer, but that should be able to be fixed with the ntdll patch.

**Comment by Mhowser** [comment]
> should be able to be fixed with the ntdll patch.

Please tell us more about this

**Comment by Mhowser** [comment]
I sent a tweet to the game Developers, https://twitter.com/c4209534/status/1259008121579671552
I would appreciate if you guys would retweet mine or make your own.

**Comment by LuFlo** [comment]
> The game seems to run fine with steam installed under wine.

I can confirm this. So now the question is, why isn't it working when using Proton? Is this a bug in Proton?

**Comment by wozeparrot** [comment]
If you use `-no-browser` when launching steam from wine. Multiplayer should also work.

**Comment by wozeparrot** [comment]
So from what I have found it seems that Scrap Mechanic 0.4.0 uses a different `steam_api.dll` which seems to be causing the problem.

**Comment by Friz64** [comment]
I have also tried getting it to run with Steam under Wine (over Lutris) but i always get stuck on the Axolot splash screen. RX 580 + Arch Linux, can anybody help me out?

**Comment by wozeparrot** [comment]
Try an older version of wine, 4.20 works for me.

**Comment by wozeparrot** [comment]
Also make sure that dxvk is enabled.

**Comment by Friz64** [comment]
I can confirm that i got it running, thank you!

**Comment by Mhowser** [comment]
@kisak-valve do you know how to find out which steam api DLL scrap mechanic is using?

**Comment by wozeparrot** [comment]
does not really seem like a different version more like some function that was compiled in for beta 0.4.0 is causing the issue.

**Comment by ghost** [comment]
Can confirm, steam in wine 4.20 method works. Tho the first time the game took almost 40 minutes to load to the main menu.

**Comment by ghost** [comment]
Wine 4.20 is not necessary. Was able to play the game with 5.7 as well, but the game takes significantly more time to load (to the main menu).
With 4.20 only the first run was extremely slow, following runs didn't take more than a few seconds. My guess is the first time the game was generating caches.

**Comment by Friz64** [comment]
I don't get the long loading times on `lutris-4.20`, `lutris-4.21`, `lutris-5.0`, `lutris-5.2` and `lutris-5.4`. For me they start with version `lutris-5.5-2`.

**Comment by Friz64** [comment]
Multiplayer works for me! Steam is just kind of weird sometimes.

**Comment by Mhowser** [comment]
I cannot get Steam running under wine, it's just so broken. Is there any progress towards getting Proton working?

**Comment by Friz64** [comment]
I just had a 2 hour gaming session (creative mode, survival mode and multiplayer), running very well.

@Mhowser Do you have Discord? I could help you with the setup. `Friz64#4876`

**Comment by Hastur3333** [comment]
I got it to work with Lutris steam under wine earlier today where I just started it and left for a while and returned to see the main menu. But now when I try it gets to the Axolot screen and then the window dissapears.
In the log it says
`X Error of failed request:  BadWindow (invalid Window parameter)
  Major opcode of failed request:  10 (X_UnmapWindow)
  Resource id in failed request:  0x1e00001
  Serial number of failed request:  93
  Current serial number in output stream:  93`

**Comment by Mhowser** [comment]
@Friz64 sent you a friend request.

**Comment by Leteen** [comment]
I'm running windows, same "Z:\Jenkins\workspace\sm\ScrapMechanic\Contraption.cpp" error

**Comment by Mubelotix** [comment]
It's working for me using Steam trough Wine thanks to Lutris. But I had to wait around 5 minutes on the "Axolot" screen and there is a lot of small freezes. But it's very playable.

**Comment by Friz64** [comment]
> But I had to wait around 5 minutes on the "Axolot" screen

Try using wine version lutris-5.4, it reduces the startup time to seconds.

> there is a lot of small freezes

Try moving the game installation to a SSD, i heard that helps with that.

**Comment by Mubelotix** [comment]
Thanks you solved the loading problem! How can my SSD fix a graphic issue?

**Comment by ghost** [comment]
I suspect the game does IO synchronously and in the main game loop, which causes it to lag a lot if the storage device is not fast enough. 
I moved the game (whole wine prefix, to be exact) to a SSD and also mount a tmpfs on the directory with the save files every time the game starts. That reduced stuttering to almost zero. (But the rest, I suspect, is on the devs of the game. There are reports of stuttering from Windows users too)

**Comment by wozeparrot** [comment]
From a proton log this line seems interesting `45061.400:0032:0033:err:steamclient:create_win_interface Don't recognize interface name: SteamNetworkingSockets008`

Right after this line the game crashes.

**Comment by ghost** [comment]
Tested on proton 5.0.7, can get to the menu, cannot load either a saved game or a new one.

**Comment by wozeparrot** [comment]
Yea its an issue with how steam connects the linux client to the game running in proton.

The game creates a server when you create or load a game using Steam's networking but clearly its not implemented yet.

**Comment by stevenlafl** [comment]
Just bought it, can't play it. If we could find a workaround that would be great. I can do anything from the screen, but as soon as I go to load/create a game it crashes to the bugsplat screen, and that entry is in the logs `18:06:38 (1) ERROR: Z:\Jenkins\workspace\sm\ScrapMechanic\Contraption.cpp:862 Bugsplat!`. I've tried all the steam options, including ESYNC, WINED3D, etc. listed here https://github.com/ValveSoftware/Proton/#runtime-config-options 

In the Proton log I see: `104920.196:00d0:00d4:err:steamclient:create_win_interface Don't recognize interface name: SteamNetworkingSockets008` that appears to be just prior to loading up Bugsplat dlls. 

A backtrace:
```
Thread 1 "00b4" received signal SIGSEGV, Segmentation fault.
0x00685305 in ?? ()
   from /media/stevenlafl/Helper/SteamLibrary/steamapps/compatdata/387990/pfx/dosdevices/z:/media/stevenlafl/Helper/SteamLibrary/steamapps/common/Scrap Mechanic/Release/ScrapMechanic.exe
```
`Wine-gdb> backtrace`
```
#0  0x00685305 in ?? ()
   from /media/stevenlafl/Helper/SteamLibrary/steamapps/compatdata/387990/pfx/dosdevices/z:/media/stevenlafl/Helper/SteamLibrary/steamapps/common/Scrap Mechanic/Release/ScrapMechanic.exe
#1  0x006bb68c in ?? ()
   from /media/stevenlafl/Helper/SteamLibrary/steamapps/compatdata/387990/pfx/dosdevices/z:/media/stevenlafl/Helper/SteamLibrary/steamapps/common/Scrap Mechanic/Release/ScrapMechanic.exe
#2  0x006b85a2 in ?? ()
   from /media/stevenlafl/Helper/SteamLibrary/steamapps/compatdata/387990/pfx/dosdevices/z:/media/stevenlafl/Helper/SteamLibrary/steamapps/common/Scrap Mechanic/Release/ScrapMechanic.exe
#3  0x00594632 in ?? ()
   from /media/stevenlafl/Helper/SteamLibrary/steamapps/compatdata/387990/pfx/dosdevices/z:/media/stevenlafl/Helper/SteamLibrary/steamapps/common/Scrap Mechanic/Release/ScrapMechanic.exe
#4  0x00592dd9 in ?? ()
   from /media/stevenlafl/Helper/SteamLibrary/steamapps/compatdata/387990/pfx/dosdevices/z:/media/stevenlafl/Helper/SteamLibrary/steamapps/common/Scrap Mechanic/Release/ScrapMechanic.exe
#5  0x00591ccc in ?? ()
   from /media/stevenlafl/Helper/SteamLibrary/steamapps/compatdata/387990/pfx/dosdevices/z:/media/stevenlafl/Helper/SteamLibrary/steamapps/common/Scrap Mechanic--Type <RET> for more, q to quit, c to continue without paging--
/Release/ScrapMechanic.exe
#6  0x006b98c9 in ?? ()
   from /media/stevenlafl/Helper/SteamLibrary/steamapps/compatdata/387990/pfx/dosdevices/z:/media/stevenlafl/Helper/SteamLibrary/steamapps/common/Scrap Mechanic/Release/ScrapMechanic.exe
#7  0x00d1ed33 in ?? ()
   from /media/stevenlafl/Helper/SteamLibrary/steamapps/compatdata/387990/pfx/dosdevices/z:/media/stevenlafl/Helper/SteamLibrary/steamapps/common/Scrap Mechanic/Release/ScrapMechanic.exe
#8  0x7b4539c2 in call_process_entry ()
   from /media/stevenlafl/Helper/SteamLibrary/steamapps/compatdata/387990/pfx/dosdevices/c:/windows/syswow64/kernel32.dll
#9  0x7b453da5 in start_process ()
   from /media/stevenlafl/Helper/SteamLibrary/steamapps/compatdata/387990/pfx/dosdevices/c:/windows/syswow64/kernel32.dll
#10 0x7b4539ce in __wine_start_process ()
   from /media/stevenlafl/Helper/SteamLibrary/steamapps/compatdata/387990/pfx/dosdevices/c:/windows/syswow64/kernel32.dll
#11 0x00000000 in ?? ()
```

If there were some better explanation of what this line is, and what it's trying to do, perhaps one of us could find a way to solve it.

A similar note on a different game searching for the error suggests using https://github.com/GloriousEggroll/proton-ge-custom/releases . I may try this in a bit.

**Comment by wozeparrot** [comment]
Use lutris winesteam and install the game there.

**Comment by stevenlafl** [comment]
@wozeparrot I tried Lutris Winesteam, and it did not work at first. The most recent version I saw that worked with it was `lutris-fshack-5.7-x86_64`. 5.6 and 5.7 itself crash before it hit the menu.

Regular Wine or Wine Staging keeps reloading the Steam Library over and over so I can't click on anything.

I'm not sure what is different in "fshack" version, but it works. It takes a very, very long time to load (~5 min for menu, ~5 min for game) - but once it does - play seems to be smooth as glass except for a few VERY small stutters when loading assets.

Others have mentioned 5.4 may work to get these load times down to something faster, I might try that too.

Edit: Confirmed **lutris-fshack-5.4** is the best answer. Fast as lightning loads.
Edit 2-3: It looks like the problem is that sdk v. 148 is not implemented @kisak-valve . I've been looking at the differences and trying to compile another version of Proton based on `proton-4.12-13`  (last known fast-loading proton packaged with Steam Debian) that contains the SteamNetworkingSockets008 interface. If it's similar to SteamNetworkingSockets006 (I saw there were ~5 new methods including Certificates and Polls) then maybe it will work. There are some data structures that may malign if the difference is too large. Obviously since you guys have the code I'm sure you're working on it already.

**Comment by wozeparrot** [comment]
Yea there is a cw label which im assuming is currently working

**Comment by Mhowser** [comment]
That currently working label shouldn't apply here because it technically still doesn't work with a standard Steam/Proton setup. It should be changed to workaround or something.

**Comment by TaiTair** [comment]
> That currently working label shouldn't apply here because it technically still doesn't work with a standard Steam/Proton setup. It should be changed to workaround or something.

Agreed. In the context of Proton, running steam in a wineprefix with wine is not what I would call working. Current solution doesn't even use Proton.

**Comment by stevenlafl** [comment]
OK, I've implemented the SteamNetworkingSockets008 interface and the game runs. While I have a PR open for 5.0 branch, I chose to compile against 4.11-13 for the reasons above: it loads fastest. Here's a link to that package **https://github.com/stevenlafl/Proton/releases/tag/proton-4.11-13-sdk-148**

Cheers!

Edit: Back link to PR https://github.com/ValveSoftware/Proton/pull/3880
Edit: if you need the latest proton (5.0-7) you can grab it here: https://github.com/stevenlafl/Proton/releases/tag/proton-5.0-7-sdk148

**Comment by ghost** [comment]
> OK, I've implemented the SteamNetworkingSockets008 interface and the game runs. While I have a PR open for 5.0 branch, I chose to compile against 4.11-13 for the reasons above: it loads fastest. Here's a link to that package https://github.com/stevenlafl/Proton/releases/tag/proton-4.11-13-sdk-148
> 
> Cheers!

I hope Your PR will make it to the next Proton update

**Comment by TaiTair** [comment]
> OK, I've implemented the SteamNetworkingSockets008 interface and the game runs. While I have a PR open for 5.0 branch, I chose to compile against 4.11-13 for the reasons above: it loads fastest. Here's a link to that package https://github.com/stevenlafl/Proton/releases/tag/proton-4.11-13-sdk-148
> 
> Cheers!

Compiled it against latest proton and it works great! Thanks @stevenlafl great work!

**Comment by stevenlafl** [comment]
> > OK, I've implemented the SteamNetworkingSockets008 interface and the game runs. While I have a PR open for 5.0 branch, I chose to compile against 4.11-13 for the reasons above: it loads fastest. Here's a link to that package https://github.com/stevenlafl/Proton/releases/tag/proton-4.11-13-sdk-148
> > Cheers!
> 
> Compiled it against latest proton and it works great! Thanks @stevenlafl great work!

Would you mind reapplying the changes and trying again? Accidentally committed 4.11-13's git submodules.

**Comment by TaiTair** [comment]
Ah yeah I just grabbed the latest proton git with its submodules and then I applied your PR(#3880) with the two commits to it. I'm sure it'll be fine.

**Comment by gtasseron** [comment]
> > OK, I've implemented the SteamNetworkingSockets008 interface and the game runs. While I have a PR open for 5.0 branch, I chose to compile against 4.11-13 for the reasons above: it loads fastest. Here's a link to that package https://github.com/stevenlafl/Proton/releases/tag/proton-4.11-13-sdk-148
> > Cheers!
> 
> Compiled it against latest proton and it works great! Thanks @stevenlafl great work!

Is it possible that you provide us with the compiled package?

**Comment by stevenlafl** [comment]
> > > OK, I've implemented the SteamNetworkingSockets008 interface and the game runs. While I have a PR open for 5.0 branch, I chose to compile against 4.11-13 for the reasons above: it loads fastest. Here's a link to that package https://github.com/stevenlafl/Proton/releases/tag/proton-4.11-13-sdk-148
> > > Cheers!
> > 
> > 
> > Compiled it against latest proton and it works great! Thanks @stevenlafl great work!
> 
> Is it possible that you provide us with the compiled package?

For 5.0-7? Yeah, I just recompiled it for proton latest (5.0-7) and edited my post above.

**Comment by AnasR7** [comment]
> > > > OK, I've implemented the SteamNetworkingSockets008 interface and the game runs. While I have a PR open for 5.0 branch, I chose to compile against 4.11-13 for the reasons above: it loads fastest. Here's a link to that package https://github.com/stevenlafl/Proton/releases/tag/proton-4.11-13-sdk-148
> > > > Cheers!
> > > 
> > > 
> > > Compiled it against latest proton and it works great! Thanks @stevenlafl great work!
> > 
> > 
> > Is it possible that you provide us with the compiled package?
> 
> For 5.0-7? Yeah, I just recompiled it for proton latest (5.0-7) and edited my post above.


great patch  @stevenlafl , btw im using your patch 4.11 and 5.0-7 and its working but i have problem with lag every time i play, seems to be same problem with this dude in protondb page
[Link Protondb Page](https://www.protondb.com/app/387990#arSP-RXVHn)

![image](https://user-images.githubusercontent.com/23519166/83080004-4670c680-a0a7-11ea-8546-8892026caecd.png)

im using
``` 
- Laptop ASUS TUF FX505DY
- manjaro OS KDE
- RAM 8GB
- CPU AMD Ryzen 5
- GPU AMD ATI Radeon RX 560X
```

BTW GREAT patch.

**Comment by Mhowser** [comment]
It might be compiling shaders on the fly

**Comment by Friz64** [comment]
It's not normal for it to load textures that long in my experience. Besides moving the game to an SSD, i also heard that moving it off of a NTFS partition to something like EXT4 helps.

**Comment by Mubelotix** [comment]
But lags are not a Linux related problem. Especially on multiplayer. Multiplayer is not playable.

**Comment by stevenlafl** [comment]
I agree. I have an NTFS spin-disk HDD drive I run this off of on Linux. My main reason for playing this is multiplayer, so that was my focus. It will spike lagging for 5 minutes for me, tops, and then be fine from then forward. I have maxed out all settings. Multiplayer works great for me.

I will note that there does since appear to have been an update to the game. I have not tested these yet.

**Comment by gordominossi** [comment]
It's not working for me. I tried running it with both of your patches, @stevenlafl, and I get a black screen for less then a second and then the game closes. Is there a place where I can find some logs or anything that provides more information so I can fix this?

**Comment by gordominossi** [comment]
Found logs under Scrap Mechanic/logs but it isn't of much help
```15:18:38 (1) [Default] Initialized Logger
15:18:38 (1) [System] Operating system: Win 7 Service pack 1
15:18:38 (1) [System] Permissions for C:/users/steamuser/Application Data - Others write | Others read | Group write | Group read | Owner write | Owner read
15:18:38 (1) [System] Permissions for C:/users/steamuser/Local Settings/Application Data - Others write | Others read | Group write | Group read | Owner write | Owner read
15:18:38 (1) [System] Permissions for .. - Others write | Others read | Group write | Group read | Owner write | Owner read
15:18:38 (1) WARNING: Z:\Jenkins\workspace\sm\ScrapMechanic\GameSettings.cpp:66 Settings file not found: 'C:/users/steamuser/Application Data/Axolot Games/Scrap Mechanic/User/User_76561198251504110/settings.json'
15:18:38 (1) WARNING: Z:\Jenkins\workspace\sm\ScrapMechanic\GameSettings.cpp:82 Settings value missing! (VerticalSync). Defaulting to: 0
15:18:38 (1) WARNING: Z:\Jenkins\workspace\sm\ScrapMechanic\GameSettings.cpp:82 Settings value missing! (FOV). Defaulting to: 1
15:18:38 (1) WARNING: Z:\Jenkins\workspace\sm\ScrapMechanic\GameSettings.cpp:82 Settings value missing! (InvYAxisState). Defaulting to: 0
15:18:38 (1) WARNING: Z:\Jenkins\workspace\sm\ScrapMechanic\GameSettings.cpp:82 Settings value missing! (VehicleCameraMode). Defaulting to: 1
15:18:38 (1) WARNING: Z:\Jenkins\workspace\sm\ScrapMechanic\GameSettings.cpp:96 Settings value missing! (FrameRateCap). Defaulting to: 0
15:18:38 (1) WARNING: Z:\Jenkins\workspace\sm\ScrapMechanic\GameSettings.cpp:96 Settings value missing! (MouseSpeed). Defaulting to: 0.5
15:18:38 (1) WARNING: Z:\Jenkins\workspace\sm\ScrapMechanic\GameSettings.cpp:96 Settings value missing! (MasterVolume). Defaulting to: 1
15:18:38 (1) WARNING: Z:\Jenkins\workspace\sm\ScrapMechanic\GameSettings.cpp:96 Settings value missing! (AmbientVolume). Defaulting to: 0.8
15:18:38 (1) WARNING: Z:\Jenkins\workspace\sm\ScrapMechanic\GameSettings.cpp:96 Settings value missing! (MusicVolume). Defaulting to: 0.8
15:18:38 (1) WARNING: Z:\Jenkins\workspace\sm\ScrapMechanic\GameSettings.cpp:96 Settings value missing! (EffectVolume). Defaulting to: 0.8
15:18:38 (1) WARNING: Z:\Jenkins\workspace\sm\ScrapMechanic\GameSettings.cpp:96 Settings value missing! (GUIVolume). Defaulting to: 0.8
15:18:38 (1) [Default] DisplayOptions::applySettings (1920x1080) mode:3 try fullscreen: no
15:18:38 (1) [System] GPU: Intel(R) HD Graphics 630
15:18:38 (1) [System] Dedicated Video Memory: 3072 Mb
15:18:38 (1) [System] Shared System Memory: 4095 Mb
15:18:38 (1) [System] CPU: Intel(R) Core(TM) i7-7700HQ CPU @ 2.80GHz
15:18:38 (1) [System] Total RAM: 15777 Mb
15:18:38 (1) [System] Available RAM: 13935 Mb
15:18:38 (1) [System] Total Virtual RAM: 4095 Mb
15:18:38 (1) [System] Available Virtual RAM: 4095 Mb

**Comment by Mubelotix** [comment]
@gordominossi looks like it's a different issue. Have you tried running the game *and* Steam in wine with Lutris?

**Comment by gordominossi** [comment]
@Mubelotix, I've tried to run it under Lutris as well but I can't find out how to set the wine version to a custom executable, so I'm running it with the default wine version (lutris-5.6-2-x86_64) and I can't find logs for the last tries, but I found some old ones (from yesterday):
```20:35:17 (1) [Default] Initialized Logger
20:35:17 (1) [System] Operating system: Win 8
20:35:17 (1) [System] Permissions for C:/users/gordo/Application Data - Others write | Others read | Group write | Group read | Owner write | Owner read
20:35:17 (1) [System] Permissions for C:/users/gordo/Local Settings/Application Data - Others write | Others read | Group write | Group read | Owner write | Owner read
20:35:17 (1) [System] Permissions for .. - Others write | Others read | Group write | Group read | Owner write | Owner read
20:35:17 (1) WARNING: Z:\Jenkins\workspace\sm\ScrapMechanic\GameSettings.cpp:66 Settings file not found: 'C:/users/gordo/Application Data/Axolot Games/Scrap Mechanic/User/User_76561198251504110/settings.json'
20:35:17 (1) WARNING: Z:\Jenkins\workspace\sm\ScrapMechanic\GameSettings.cpp:82 Settings value missing! (VerticalSync). Defaulting to: 0
20:35:17 (1) WARNING: Z:\Jenkins\workspace\sm\ScrapMechanic\GameSettings.cpp:82 Settings value missing! (FOV). Defaulting to: 1
20:35:17 (1) WARNING: Z:\Jenkins\workspace\sm\ScrapMechanic\GameSettings.cpp:82 Settings value missing! (InvYAxisState). Defaulting to: 0
20:35:17 (1) WARNING: Z:\Jenkins\workspace\sm\ScrapMechanic\GameSettings.cpp:82 Settings value missing! (VehicleCameraMode). Defaulting to: 1
20:35:17 (1) WARNING: Z:\Jenkins\workspace\sm\ScrapMechanic\GameSettings.cpp:96 Settings value missing! (FrameRateCap). Defaulting to: 0
20:35:17 (1) WARNING: Z:\Jenkins\workspace\sm\ScrapMechanic\GameSettings.cpp:96 Settings value missing! (MouseSpeed). Defaulting to: 0.5
20:35:17 (1) WARNING: Z:\Jenkins\workspace\sm\ScrapMechanic\GameSettings.cpp:96 Settings value missing! (MasterVolume). Defaulting to: 1
20:35:17 (1) WARNING: Z:\Jenkins\workspace\sm\ScrapMechanic\GameSettings.cpp:96 Settings value missing! (AmbientVolume). Defaulting to: 0.8
20:35:17 (1) WARNING: Z:\Jenkins\workspace\sm\ScrapMechanic\GameSettings.cpp:96 Settings value missing! (MusicVolume). Defaulting to: 0.8
20:35:17 (1) WARNING: Z:\Jenkins\workspace\sm\ScrapMechanic\GameSettings.cpp:96 Settings value missing! (EffectVolume). Defaulting to: 0.8
20:35:17 (1) WARNING: Z:\Jenkins\workspace\sm\ScrapMechanic\GameSettings.cpp:96 Settings value missing! (GUIVolume). Defaulting to: 0.8
20:35:17 (1) [Default] DisplayOptions::applySettings (1920x1080) mode:3 try fullscreen: no
20:35:17 (1) [System] GPU: Intel(R) HD Graphics 630
20:35:17 (1) [System] Dedicated Video Memory: 3072 Mb
20:35:17 (1) [System] Shared System Memory: 4095 Mb
20:35:17 (1) [System] CPU: Intel(R) Core(TM) i7-7700HQ CPU @ 2.80GHz
20:35:17 (1) [System] Total RAM: 42383 Mb
20:35:17 (1) [System] Available RAM: 10153 Mb
20:35:17 (1) [System] Total Virtual RAM: 4095 Mb
20:35:17 (1) [System] Available Virtual RAM: 4095 Mb
20:35:17 (1) [System] Supported feature level: Unknown
20:35:17 (1) ERROR: Z:\Jenkins\workspace\sm\ContraptionCommon\renderer\dx11\d3d11_device.cpp:298 DirectX Error "D3D11CreateDeviceAndSwapChain" failed. Error code: 0x80004005

20:35:17 (1) [Render] ERROR: Z:\Jenkins\workspace\sm\ContraptionCommon\renderer\dx11\d3d11_renderer.cpp:70 Device creation failed

**Comment by gordominossi** [comment]
Hey, I got it fixed. The issues were Optimus related. Uninstalled bumblebee and installed PRIME. Everything works just fine now. I'm getting from 30 to 90 FPS on max settings. Thank you @stevenlafl for the great patch :)

**Comment by AnasR7** [comment]
With Proton 5.0-8 still got a lag in a second with amd GPU

**Comment by rmburg** [comment]
The crash when loading a world has been fixed with the latest patch.

**Comment by LukasDoesDev** [comment]
> It's not working for me. I tried running it with both of your patches, @stevenlafl, and I get a black screen for less then a second and then the game closes. Is there a place where I can find some logs or anything that provides more information so I can fix this?

I'm getting the same.

```
19:20:14 (1/0) [Default] Initialized Logger
19:20:14 (1/0) [System] Operating system: Win 8
19:20:14 (1/0) [System] Permissions for C:/users/steamuser/Application Data - Others write | Others read | Group write | Group read | Owner write | Owner read
19:20:14 (1/0) [System] Permissions for C:/users/steamuser/Local Settings/Application Data - Others write | Others read | Group write | Group read | Owner write | Owner read
19:20:14 (1/0) [System] Permissions for .. - Others write | Others read | Group write | Group read | Owner write | Owner read
19:20:14 (1/0) WARNING: Z:\Jenkins\workspace\sm\ScrapMechanic\GameSettings.cpp:68 Settings file not found: 'C:/users/steamuser/Application Data/Axolot Games/Scrap Mechanic/User/User_76561199045653350/settings.json'
19:20:14 (1/0) WARNING: Z:\Jenkins\workspace\sm\ScrapMechanic\GameSettings.cpp:84 Settings value missing! (VerticalSync). Defaulting to: 0
19:20:14 (1/0) WARNING: Z:\Jenkins\workspace\sm\ScrapMechanic\GameSettings.cpp:84 Settings value missing! (FOV). Defaulting to: 1
19:20:14 (1/0) WARNING: Z:\Jenkins\workspace\sm\ScrapMechanic\GameSettings.cpp:84 Settings value missing! (InvYAxisState). Defaulting to: 0
19:20:14 (1/0) WARNING: Z:\Jenkins\workspace\sm\ScrapMechanic\GameSettings.cpp:84 Settings value missing! (VehicleCameraMode). Defaulting to: 1
19:20:14 (1/0) WARNING: Z:\Jenkins\workspace\sm\ScrapMechanic\GameSettings.cpp:98 Settings value missing! (FrameRateCap). Defaulting to: 0
19:20:14 (1/0) WARNING: Z:\Jenkins\workspace\sm\ScrapMechanic\GameSettings.cpp:98 Settings value missing! (MouseSpeed). Defaulting to: 0.5
19:20:14 (1/0) WARNING: Z:\Jenkins\workspace\sm\ScrapMechanic\GameSettings.cpp:98 Settings value missing! (MasterVolume). Defaulting to: 1
19:20:14 (1/0) WARNING: Z:\Jenkins\workspace\sm\ScrapMechanic\GameSettings.cpp:98 Settings value missing! (AmbientVolume). Defaulting to: 0.8
19:20:14 (1/0) WARNING: Z:\Jenkins\workspace\sm\ScrapMechanic\GameSettings.cpp:98 Settings value missing! (MusicVolume). Defaulting to: 0.8
19:20:14 (1/0) WARNING: Z:\Jenkins\workspace\sm\ScrapMechanic\GameSettings.cpp:98 Settings value missing! (EffectVolume). Defaulting to: 0.8
19:20:14 (1/0) WARNING: Z:\Jenkins\workspace\sm\ScrapMechanic\GameSettings.cpp:98 Settings value missing! (GUIVolume). Defaulting to: 0.8
19:20:14 (1/0) [Default] DisplayOptions::applySettings (1680x1050) mode:3 try fullscreen: no
19:20:14 (1/0) [System] GPU: AMD Radeon HD 7700 Series
19:20:14 (1/0) [System] Dedicated Video Memory: 1024 Mb
19:20:14 (1/0) [System] Shared System Memory: 4095 Mb
19:20:14 (1/0) [System] CPU: AMD Ryzen 5 3600 6-Core Processor              
19:20:14 (1/0) [System] Total RAM: 20108 Mb
19:20:14 (1/0) [System] Available RAM: 9030 Mb
19:20:14 (1/0) [System] Total Virtual RAM: 4095 Mb
19:20:14 (1/0) [System] Available Virtual RAM: 4095 Mb
```
Using Steam Proton version 5.13-4
OS: Pop!_OS 20.10 x86_64
Kernel: 5.8.0-7630-generic
CPU: AMD Ryzen 5 3600 (12) @ 3.600GHz
GPU: AMD ATI Radeon HD 7770/8760 / R7

**Comment by LukasDoesDev** [comment]
I opened Steam in Lutris Wine, downloaded Scrap Mechanic with super slow speeds, and it launched.
UPDATE: It downloaded some workshop item for a moment and then my system crashed.

**Comment by lanodan** [comment]
Here are the results on my setup:
- System: Artix(Arch without systemd) in a chroot, host is gentoo
- CPU: AMD Ryzen 7 3700X
- GPU: AMD RX 5600 XT (mesa 20.2.6 on host, 20.3.4 in chroot)
- Kernel: 5.4.66-gentoo
- Proton: 5.0 (5.13 didn't load because bwrap doesn't seems to work in chroots)

Issues:
- Only in Sway (1.5.1, wlroots 0.12.0): Virtual desktop or not, cursor moves are broken, only click works
- With the virtual desktop I had to disable my second monitor (both in Sway and Xorg) to not have the game's window being off-center
- I had `~/.asoundrc` set for my Valve Index so as expected sound wasn't working

**Comment by LukasDoesDev** [comment]
I moved to Arch Linux and still getting same error as before
\-\-\-
Lukas
<sub>i use arch btw</sub>

**Comment by LukasDoesDev** [comment]
Turns out my driver probably didn't have Vulkan support. I used `PROTON_USE_WINED3D=1 %command%` as my launch option and it worked. Thanks to this tweet by `Fritz64_`: https://twitter.com/Friz64_/status/1370867612058578952

**Comment by kisak-valve** [comment]
Hello @LukasDoesDev, it looks like you have a Southern Islands (SI) generation AMD video card. The card does support Vulkan, but Linux uses the `radeon` kernel module by default with that generation and the `radeon` kernel module does not support Vulkan. Please give https://github.com/ValveSoftware/Proton/wiki/For-AMD-users-having-issues-with-non-OpenGL-games a read.

**Comment by Tofixrs** [comment]
Since proton 8.0 the game fails to open the debug console (enabled with the -dev flag) with the following error
![image](https://github.com/ValveSoftware/Proton/assets/73693639/c4728f3c-3fe1-42d9-bec4-863446efd7c3)
While its not necessery for normal gameplay its basically a necessity for modders same happens with the game mod tool and with the mod tool its worse due to it not launching at all (technically it launches but failes to find the game directory making it close) on versions before proton 8 but on proton 8 the console is broken

**Comment by kisak-valve** [comment]
Hello @Tofixrs, please add [`PROTON_LOG=1 %command%`](https://github.com/ValveSoftware/Proton#runtime-config-options) to the game's launch options and attach the generated $HOME/steam-$APPID.log to this issue report as a file. (Proton logs compress well if needed.) If this is a regression, please explicitly note the newest Proton version you know is not affected.

**Comment by Tofixrs** [comment]
@kisak-valve There you go
[steam-387990.log](https://github.com/ValveSoftware/Proton/files/11627889/steam-387990.log)
The newest not affected version not affected is Proton 7.0-6

**Comment by alasky17** [comment]
@Tofixrs Thank you for the report!  This regression has been fixed in Proton Experimental :)
